server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -239,15 +246,45 @@ public InternalEngine(EngineConfig engineConfig) {
     }
 
     private LocalCheckpointTracker createLocalCheckpointTracker(
-        BiFunction<Long, Long, LocalCheckpointTracker> localCheckpointTrackerSupplier) throws IOException {
-        final long maxSeqNo;
-        final long localCheckpoint;
-        final SequenceNumbers.CommitInfo seqNoStats =
-            SequenceNumbers.loadSeqNoInfoFromLuceneCommit(store.readLastCommittedSegmentsInfo().userData.entrySet());
-        maxSeqNo = seqNoStats.maxSeqNo;
-        localCheckpoint = seqNoStats.localCheckpoint;
-        logger.trace("recovered maximum sequence number [{}] and local checkpoint [{}]", maxSeqNo, localCheckpoint);
-        return localCheckpointTrackerSupplier.apply(maxSeqNo, localCheckpoint);
+        BiFunction<Long, Long, LocalCheckpointTracker> localCheckpointTrackerSupplier) {
+        try {
+            final SequenceNumbers.CommitInfo seqNoStats =
+                SequenceNumbers.loadSeqNoInfoFromLuceneCommit(store.readLastCommittedSegmentsInfo().userData.entrySet());
+            final long maxSeqNo = seqNoStats.maxSeqNo;
+            final long localCheckpoint = seqNoStats.localCheckpoint;
+            logger.trace("recovered maximum sequence number [{}] and local checkpoint [{}]", maxSeqNo, localCheckpoint);
+            final LocalCheckpointTracker tracker = localCheckpointTrackerSupplier.apply(maxSeqNo, localCheckpoint);
+            // scans existing sequence numbers in Lucene commit, then marks them as completed in the tracker.
+            if (localCheckpoint < maxSeqNo && softDeleteEnabled) {
+                try (Searcher engineSearcher = acquireSearcher("build_local_checkpoint_tracker", SearcherScope.INTERNAL)) {
+                    final DirectoryReader reader = Lucene.wrapAllDocsLive(engineSearcher.getDirectoryReader());
+                    final IndexSearcher searcher = new IndexSearcher(reader);
+                    searcher.setQueryCache(null);
+                    final Query query = LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, localCheckpoint + 1, maxSeqNo);
+                    for (LeafReaderContext leaf : reader.leaves()) {
+                        final Scorer scorer = searcher.createWeight(query, ScoreMode.COMPLETE_NO_SCORES, 1.0f).scorer(leaf);
+                        if (scorer == null) {
+                            continue;
+                        }
+                        final DocIdSetIterator docIdSetIterator = scorer.iterator();
+                        final NumericDocValues seqNoDocValues = leaf.reader().getNumericDocValues(SeqNoFieldMapper.NAME);
+                        int docId;
+                        while ((docId = docIdSetIterator.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
```

Comment:
This takes another approach than `LuceneChangesSnapshot`. @jimczi @s1monw can you please double check this?

Reply:
It should be possible to use the reader's `PointsValue` directly to speed up the search since they are always indexed.
Something like the following untested snippet:
````
void markAsCompleted(IndexReader reader, LocalCheckpointTracker tracker, long minSeqNo, long maxSeqNo) throws IOException {
        for (LeafReaderContext ctx : reader.leaves()) {
            PointValues values = ctx.reader().getPointValues(SeqNoFieldMapper.NAME);
            byte[] lowerPoint = new byte[values.getBytesPerDimension()];
            LongPoint.encodeDimension(minSeqNo, lowerPoint, 0);
            byte[] upperPoint = new byte[values.getBytesPerDimension()];
            LongPoint.encodeDimension(maxSeqNo, upperPoint, 0);
            final Bits liveDocs = ctx.reader().getLiveDocs() == null ?
                new Bits.MatchAllBits(reader.maxDoc()) : ctx.reader().getLiveDocs();
            int numBytes = values.getBytesPerDimension();
            values.intersect(new PointValues.IntersectVisitor() {
                @Override
                public void visit(int docID) throws IOException {
                    throw new IllegalStateException("should never be called");
                }

                @Override
                public void visit(int docID, byte[] packedValue) throws IOException {
                    if (liveDocs.get(docID) == false) {
                        return;
                    }

                    if (FutureArrays.compareUnsigned(packedValue, 0, numBytes, lowerPoint, 0, numBytes) < 0) {
                        // Doc's value is too low, in this dimension
                        return;
                    }
                    if (FutureArrays.compareUnsigned(packedValue, 0, numBytes, upperPoint, 0, numBytes) > 0) {
                        // Doc's value is too high, in this dimension
                        return;
                    }
                    long seqNo = LongPoint.decodeDimension(packedValue, 0);
                    tracker.markSeqNoAsCompleted(seqNo);
                }

                @Override
                public PointValues.Relation compare(byte[] minPackedValue, byte[] maxPackedValue) {
                    if (FutureArrays.compareUnsigned(minPackedValue, 0, numBytes, upperPoint, 0, numBytes) > 0 ||
                            FutureArrays.compareUnsigned(maxPackedValue, 0, numBytes, lowerPoint, 0, numBytes) < 0) {
                        return PointValues.Relation.CELL_OUTSIDE_QUERY;
                    }
                    return PointValues.Relation.CELL_CROSSES_QUERY;
                }
            });

        }
    }
````
This would avoid the need to execute a query to read from doc values but I am not sure if it's worth the complexity. If only few documents should be visited this is probably useless and using a query is the right approach.

Reply:
I pull this in, and it works perfectly. However, I am not sure if we should use it either. Normally we would expect to visit a few to several hundred documents.

Reply:
I agree this is probably a premature optimization. I wrote it mainly to see how it looks but this shouldn't bring much if the number of documents to visit is low. The query+doc_values approach seems easier to maintain.